lib/proposed_api/cdk-pipeline.ts

```diff
@@ -50,6 +71,76 @@ export class CdkPipeline extends cdk.Construct {
   }
 
   public addStage(stageOptions: codepipeline.StageOptions): codepipeline.IStage {
-    return this.pipeline.addStage(stageOptions);
+    class WrappedStage implements codepipeline.IStage {
+      public readonly actions: codepipeline.IAction[];
+      public readonly pipeline: codepipeline.IPipeline;
+      public readonly stageName: string;
+
+      public constructor(private readonly stage: codepipeline.IStage, private readonly stackStage: StackStage) {
+        this.actions = stage.actions;
+        this.pipeline = stage.pipeline;
+        this.stageName = stage.stageName;
+      }
+
+      public onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps): events.Rule {
+        return this.stage.onStateChange(name, target, options);
+      }
+
+      public addAction(action: codepipeline.IAction): void {
+        this.stage.addAction(action);
+        this.stackStage.addActionIfDeploysStack(action);
+      }
+    }
+
+    // ToDo handle stages inserted into arbitrary positions
+    const latestStackStage = new StackStage();
+    this.stackStages.push(latestStackStage);
```

Comment:
Given that an `IStage` exposes its `IActions`, can we not get a giant flat list of all DeployActions out (sorted by stage and then by runorder) and check against those?

Reply:
Yeah, that's probably a simpler way to do this 🤪

Reply:
Oh no, now I remember why I did it this way.

`IStage.actions` will only return the lower-level Actions (`CloudFormationCreateReplaceChangeSetAction` and `CloudFormationExecuteChangeSetAction`), which don't have a reference to the `cdk.Stack` that is being deployed. So we need to do this this way to not lose the reference to `cdk.Stack`.

Reply:
Can we not make the StackDeployAction `extends` a ExecuteChangeSetAction?

Reply:
See my comment here: https://github.com/NetaNir/meerkats/pull/8/files#diff-24560c0dec84607ac34a87450140112dR71

If it's so hard to define "custom" Pipeline Actions (have them bundled or what have you), then I feel we've done something wrong in the design there. Is that something we can fix?

As another example, this is also wholly unintuitive to me: https://github.com/NetaNir/meerkats/blob/master/lib/proposed_api/deploy-cdk-stack-action.ts#L71

Reply:
> See my comment here: https://github.com/NetaNir/meerkats/pull/8/files#diff-24560c0dec84607ac34a87450140112dR71

This one should be simple, I think:

```ts
export class PublishAssetsAction implements codepipeline.IAction {
  public readonly actionProperties: codepipeline.ActionProperties;
  private action: codepipeline.IAction;

  constructor(private readonly props: PublishAssetsActionProps) {
    this.actionProperties = {
      actionName: props.actionName,
      artifactBounds: { minInputs: 1, maxInputs: 5, minOutputs: 0, maxOutputs: 5 },
      category: codepipeline.ActionCategory.BUILD,
      provider: 'CodeBuild',
    };
  }

  public bind(scope: cdk.Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
      codepipeline.ActionConfig {
    const project = new codebuild.PipelineProject(scope, 'Default', {
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          install: {
            commands: `(curl -o dl.zip -L "${this.props.vendoredGitHubLocation}" && unzip dl.zip && cd ${this.props.vendorZipDir} && npm install -g *.tgz)`,
          },
          build: {
            commands: 'for manifest in *.assets.json; do cdk-assets --path $manifest --verbose publish; done',
          },
        },
      }),
    });

    project.addToRolePolicy(new iam.PolicyStatement({
      actions: ['sts:AssumeRole'],
      resources: ['arn:*:iam::*:role/*-publishing-role-*'],
    }));

    this.action = new codepipeline_actions.CodeBuildAction({
      actionName: 'Publish',
      project,
      input: this.props.cloudAssemblyInput,
    });

    return this.action.bind(scope, stage, options);
  }

  public onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps): events.Rule {
    return this.action.onStateChange(name, target, options);
  }
}
```

Reply:
> As another example, this is also wholly unintuitive to me: https://github.com/NetaNir/meerkats/blob/master/lib/proposed_api/deploy-cdk-stack-action.ts#L71

Why? How else would you express composite (expanding to multiple ones) Actions?

Reply:
> Oh no, now I remember why I did it this way.
> 
> `IStage.actions` will only return the lower-level Actions (`CloudFormationCreateReplaceChangeSetAction` and `CloudFormationExecuteChangeSetAction`), which don't have a reference to the `cdk.Stack` that is being deployed. So we need to do this this way to not lose the reference to `cdk.Stack`.

Actually, I'm wrong again 🤪There is the `CloudFormationCreateReplaceChangeSetAction` in `Stage.actions`, but there indeed is a `DeployCdkStackAction` there as well.

Will try to change to this way, and see if it is simpler.

Reply:
> This one should be simple, I think:

Yes but you'll have to agree this is not very nice. Before, I used to only have to know that "I" (the PublishAssetAction) wanted to expand into a CodeBuildAction. Now all of a sudden I have to know how about the internals of the CodeBuildAction are set up, breaking the abstraction.

> Why? How else would you express composite (expanding to multiple ones) Actions?

I'm not sure. Maybe there is an `addToPipeline()` method and normally actions do `pipeline.addAction(this)`, and the compound action can do `pipeline.addAction(inner1); pipeline.addAction(inner2);`. Or something. What I'm objecting to is doing an `addAction(); bind()` in response to a `bind()`. It requires some deep knowledge of how the library is implemented, and to me at least is not obvious that this is the right way.

I get that this is what we have to do *today*, but I'm asking if you just want to accept this as a given, or whether you think there's a way to change the object interactions to make this all behave more logically, and make it more easily composable. I get that you probably don't see this as a problem because you *have* the aforementioned "deep knowledge of the library", and you know that *obviously this call follows this call and then you do this other thing, and also you can "just" copy/paste the actionProperties from this other class. But please take a couple of minutes to step back and try to look at the code without preconceptions of how it "should" work, pretend you know nothing about the CodePipeline library, and see if you still find it logical.

Reply:
> > This one should be simple, I think:
> 
> Yes but you'll have to agree this is not very nice. Before, I used to only have to know that "I" (the `PublishAssetAction`) wanted to expand into a `CodeBuildAction`. Now all of a sudden I have to know how about the internals of the `CodeBuildAction` are set up, breaking the abstraction. 

So, this is interesting. What if you could provide some of the things that are currently in `ActionProperties` through `bind`?

In other words, if `PublishAssetsAction` looked like this:

```ts
export class PublishAssetsAction implements codepipeline.IAction {
  public readonly actionProperties: codepipeline.ActionProperties;
  private action: codepipeline.IAction;

  constructor(private readonly props: PublishAssetsActionProps) {
    this.actionProperties = {
      actionName: 'Publish',
    };
  }

  public bind(scope: cdk.Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
      codepipeline.ActionConfig {
    const project = new codebuild.PipelineProject(scope, 'Default', {
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          install: {
            commands: `(curl -o dl.zip -L "${this.props.vendoredGitHubLocation}" && unzip dl.zip && cd ${this.props.vendorZipDir} && npm install -g *.tgz)`,
          },
          build: {
            commands: 'for manifest in *.assets.json; do cdk-assets --path $manifest --verbose publish; done',
          },
        },
      }),
    });

    project.addToRolePolicy(new iam.PolicyStatement({
      actions: ['sts:AssumeRole'],
      resources: ['arn:*:iam::*:role/*-publishing-role-*'],
    }));

    this.action = new codepipeline_actions.CodeBuildAction({
      actionName: 'Publish',
      project,
      input: this.props.cloudAssemblyInput,
    });

    return this.action.bind(scope, stage, options);
  }

  public onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps): events.Rule {
    return this.action.onStateChange(name, target, options);
  }
}
```

Would you consider this good enough, or still missing something?

Reply:
> > Why? How else would you express composite (expanding to multiple ones) Actions?
> 
> I'm not sure. Maybe there is an `addToPipeline()` method and normally actions do `pipeline.addAction(this)`, and the compound action can do `pipeline.addAction(inner1); pipeline.addAction(inner2);`. Or something. What I'm objecting to is doing an `addAction(); bind()` in response to a `bind()`. It requires some deep knowledge of how the library is implemented, and to me at least is not obvious that this is the right way.
> 
> I get that this is what we have to do _today_, but I'm asking if you just want to accept this as a given, or whether you think there's a way to change the object interactions to make this all behave more logically, and make it more easily composable. I get that you probably don't see this as a problem because you _have_ the aforementioned "deep knowledge of the library", and you know that *obviously this call follows this call and then you do this other thing, and also you can "just" copy/paste the actionProperties from this other class. But please take a couple of minutes to step back and try to look at the code without preconceptions of how it "should" work, pretend you know nothing about the CodePipeline library, and see if you still find it logical.

I don't get your argument. Why is that `pipeline.addAction(this)` and `pipeline.addAction(inner2)` is obvious, and doing `stage.addAction()` in `bind` is not?

Notice that the `addAction` in `bind()` is exactly the same `addAction` that is part of the public API of CodePipeline in code like this:

```ts
const stage = pipeline.addStage('MyStage');
stage.addAction(new MyAction({ ... }));
```

Also, the way the types are set up, `bind()` is the only possible place you can add an action to, simply because it's the only place that gives you access to an `IStage` that the Action is a part of.

Maybe `bind()` is the confusing name here? What if it was called `addedToPipelineCallback`? Would it make sense to you that this is the only place a composite Action can happen, considering Actions can be created like this:

```ts
// not part of a CodePipeline yet - so no sub-action can exist at this point!
const action = new MyCompositeAction({ ... });
```

?